<!-- explorer/templates/explorer/details.html -->

{% extends "explorer/base.html" %}
{% load static %}
{% block content %}
{% load custom_filters %}

<div id="domain_tooltip" class="domain_tooltip"></div>

<section class="vf-content | vf-stack vf-stack--400">
    <h2> Protein Family: {{ mgyf }} </h2>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Overview</h3>
            <table class="vf-table | vf-u-width__100">
                <tbody class="vf-table__body">
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    This is the top-scoring MGnify protein (along with its specific region if not whole)
                                    that was recruited in the family through hmmsearch. Links to the MGnify Proteins site.
                                </span>
                            </span>
                            Family representative sequence
                        </td>
                        <td class="vf-table__cell">
                            {% if region_start != "" and region_end != "" %}
                                <a href="https://www.ebi.ac.uk/metagenomics/proteins/{{ protein_rep }}/?start={{ region_start }}&end={{ region_end }}" target="_blank">{{ protein_rep }}/{{ region_start }}-{{region_end}}</a>
                            {% else %}
                                <a href="https://www.ebi.ac.uk/metagenomics/proteins/{{ protein_rep }}" target="_blank">{{ protein_rep }}</a>
                            {% endif %}
                        </td>
                    </tr>
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    # Amino Acids (AA)
                                </span>
                            </span>
                            Representative length
                        </td>
                        <td class="vf-table__cell">
                            {{ rep_length }}
                        </td>
                    </tr>
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    The total number of MGnify sequences that have been iteratively recruited in the family through a series
                                    of processes such as: creating a seed alignment from the family's initial cluster, building an HMM model,
                                    and finally recruiting and aligning sequences from MGnify Proteins with the family HMM model.
                                </span>
                            </span>
                            Total number of sequences in the family
                        </td>
                        <td class="vf-table__cell">
                            {{ full_size }}
                        </td>
                    </tr>
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    Denotes if FunFam functional annotation hits were identified via the hmmer/hmmsearch tool.
                                </span>
                            </span>
                            Sequence-HMM FunFam matches
                        </td>
                        <td class="vf-table__cell">
                            {% if funfams_data %}
                                <i class="icon icon-common icon-check-square"></i>
                            {% else %}
                                <i class="icon icon-common icon-close"></i>
                            {% endif %}
                        </td>
                    </tr>
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    Denotes if Pfam domain annotation hits were identified through model searching with the hhsuite/hhblits tool.
                                </span>
                            </span>
                            Profile-profile Pfam matches
                        </td>
                        <td class="vf-table__cell">
                            {% if hits_data %}
                                <i class="icon icon-common icon-check-square"></i>
                            {% else %}
                                <i class="icon icon-common icon-close"></i>
                            {% endif %}
                        </td>
                    </tr>
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    Denotes if structure homologs of the family's representative sequence have been identified 
                                    in the AlphaFoldDB or PDB databases through the foldseek tool.
                                </span>
                            </span>
                            Structure-structure hits
                        </td>
                        <td class="vf-table__cell">
                            {% if structural_annotations %}
                                <i class="icon icon-common icon-check-square"></i>
                            {% else %}
                                <i class="icon icon-common icon-close"></i>
                            {% endif %}
                        </td>
                    </tr>
                    {% if converged %}
                    <tr class="vf-table__row">
                        <td class="vf-table__cell">
                            <span class="tooltip">
                                <i class="fas fa-question-circle tooltip-icon"></i>
                                <span class="tooltiptext">
                                    This family has refined its model and stopped recruiting
                                    new sequences from the pool in three or less iterations.
                                </span>
                            </span>
                            Converged
                        </td>
                        <td class="vf-table__cell">
                            <i class="icon icon-common icon-check-square"></i>
                        </td>
                    </tr>
                    {% endif %}
                </tbody>
            </table>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-grid vf-grid__col-4 | vf-card__content">
            <div class="vf-grid__col--span-1">
                <h3>ESMFold structure</h3>
                <p>
                    Predicted 3D protein structure through the Meta AI ESMFold model.
                    ESMFold uses the representations from a large language model (ESM2)
                    to generate an accurate structure prediction from the sequence of a protein.
                </p>
                <p>
                    For more information visit:
                    <ul class="vf-list">
                        <li class="vf-list__item">
                            <a href="https://esmatlas.com/" target="_blank" class="vf-link ext">ESM Metagenomic Atlas</a>
                        </li>
                        <li class="vf-list__item">
                            <a href="https://www.science.org/doi/full/10.1126/science.ade2574"
                                target="_blank"
                                class="vf-link ext">
                                Evolutionary-scale prediction of atomic-level protein structure with a language model
                            </a>
                        </li>
                    </ul>
                </p>
                <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='cif_blob' %}" download="{{ mgyf }}.cif" class="vf-button vf-button--secondary vf-button--sm">Download CIF file</a>
            </div>
            <div class="vf-grid__col--span-3">
                <div id="pdb_viewer">
                    <pdbe-molstar
                        custom-data-url="{% url 'serve_blob_as_file' pk=mgyf_id column_name='cif_blob' %}"
                        custom-data-format="cif"
                        alphafold-view="true"
                        bg-color-r="255"
                        bg-color-g="255"
                        bg-color-b="255"
                        sequence-panel="true"
                        landscape="true"
                    >
                    </pdbe-molstar>
                </div>
                <div>
                    <p>
                        <span class="legendColor" style="background-color: rgb(0, 83, 214);">&nbsp;</span> Very high (pLDDT â‰¥ 90)
                        <span class="legendColor" style="background-color: rgb(101, 203, 243);">&nbsp;</span> High (90 > pLDDT â‰¥ 70)
                        <span class="legendColor" style="background-color: rgb(255, 219, 19);">&nbsp;</span> Low (70 > pLDDT â‰¥ 50)
                        <span class="legendColor" style="background-color: rgb(255, 125, 69);">&nbsp;</span> Very low (pLDDT < 50)
                    </p>
                </div>
                <div>
                    <p>
                        <span class="tooltip">
                            <i class="fas fa-question-circle tooltip-icon"></i>
                            <span class="tooltiptext">
                                pLDDT corresponds to the model's prediction of its score on the per-residue Local Distance Difference Test.
                                It is a measure of local accuracy. Confidence bands are used to colour-code the residues in the 3D viewer.
                                The exact per-residue pLDDT value is shown when you mouseover the structure.
                            </span>
                        </span>
                        Average structure plddt score: <span class="plddtColor"> {{ plddt }} </span>
                        <br />
                        <span class="tooltip">
                            <i class="fas fa-question-circle tooltip-icon"></i>
                            <span class="tooltiptext">
                                The pTM score (predicted Template Modeling score) is a confidence metric that
                                estimates how accurate the global topology of a predicted protein structure is likely to be.
                            </span>
                        </span>
                        pTM score: {{ ptm }}
                    </p>
                </div>
            </div>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Feature viewer - predicted secondary structure</h3>
            <div id="feature-data" data-url="{% url 'serve_blob_as_file' pk=mgyf_id column_name='s4pred_blob' %}"></div>
            <div id="featuresContainer"></div>
            <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='s4pred_blob' %}" download="{{ mgyf }}_features.json" class="vf-button vf-button--secondary vf-button--sm">Download features JSON file</a>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">
                Multiple Sequence Alignment (Seed)
                <span class="tooltip">
                    <i class="fas fa-question-circle tooltip-icon"></i>
                    <span class="tooltiptext">
                        The seed alignment is used to create the HMM model of the family.
                        The full alignment incorporates all sequences that have been recruited in the family
                        after searching with the HMM model against the sequence pool.
                        The full alignment is usually quite larger than the seed one.
                    </span>
                </span>
            </h3>
            <div id="msa-data" data-url="{% url 'serve_blob_as_file' pk=mgyf_id column_name='seed_msa_blob' %}"></div>
            <div id="msa-div"></div>
            <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='seed_msa_blob' %}" download="{{ mgyf }}_seed_msa.fas" class="vf-button vf-button--secondary vf-button--sm">Download seed MSA file</a>
            <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='seed_msa_blob' %}" download="{{ mgyf }}_msa.fas" class="vf-button vf-button--secondary vf-button--sm">Download full MSA file</a>
        </div>
    </article>

    {% comment %} TODO {% endcomment %}
    {% comment %} <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">HMM viewer</h3>
            <div>
                <p>
                    Stack height is the information content (aka relative entropy) of the position,
                    and letters divide that height according to their estimated probability.
                    Click on a stack to highlight the relative column on the seed MSA viewer above.
                </p>
            </div>
            <div id="logo" class="logo"></div>
            <div id="col_info" class="col_info"></div>
            <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='hmm_blob' %}" download="{{ mgyf }}.hmm" class="vf-button vf-button--secondary vf-button--sm">Download HMM file</a>
            <form action="{% url 'submit_hmmsearch' mgyf_id=mgyf_id %}" method="post" target="_blank" style="display:inline;">
                {% csrf_token %}
                <button type="submit" class="vf-button vf-button--secondary vf-button--sm">Submit HMM Search</button>
            </form>
        </div>
    </article> {% endcomment %}

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400 ">
            <h3 class="vf-card__heading">
                Biomes distribution
                <span class="tooltip">
                    <i class="fas fa-question-circle tooltip-icon"></i>
                    <span class="tooltiptext">
                        An interactive sunburst plot of the biomes where the family's underlying MGnify proteins were found in.
                    </span>
                </span>
            </h3>
            <div id="biomes-data" data-url="{% url 'serve_blob_as_file' pk=mgyf_id column_name='biome_blob' %}"></div>
            <div id="sunburst-div"></div>
            <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='biome_blob' %}" download="{{ mgyf }}_biomes.csv" class="vf-button vf-button--secondary vf-button--sm">Download biomes CSV file</a>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Domain architecture</h3>
            <div>
                <p>
                    The top 15 prevalent domain architectures (MGnifams and Pfams) of the underlying family sequences.
                    The numbers on the left indicate how many times each domain architecture appeared.
                </p>
            </div>
            <div id="domain-data" data-url="{% url 'serve_blob_as_file' pk=mgyf_id column_name='domain_blob' %}"></div>
            <div id="architecturesContainer"></div>
            <a href="{% url 'serve_blob_as_file' pk=mgyf_id column_name='domain_blob' %}" download="{{ mgyf }}_domains.json" class="vf-button vf-button--secondary vf-button--sm">Download domains JSON file</a>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Functional annotation though Funfam matches</h3>
            <div>
                <p>
                    The family representative sequence was searched against the FunFam database (ver. 4.3.0) with hmmer/hmmsearch.
                </p>
            </div>
            {% if funfams_data %}
                <table id="pfams-table" class="vf-table .vf-u-margin__bottom--400">
                    <thead class="vf-table__header">
                        <tr class="vf-table__row">
                            <th class="vf-table__heading" scope="col">FunFam</th>
                            <th class="vf-table__heading" scope="col">E-value</th>
                            <th class="vf-table__heading" scope="col">Score</th>
                            <th class="vf-table__heading" scope="col">HMM from</th>
                            <th class="vf-table__heading" scope="col">HMM to</th>
                            <th class="vf-table__heading" scope="col">Alignment from</th>
                            <th class="vf-table__heading" scope="col">Alignment to</th>
                            <th class="vf-table__heading" scope="col">Envelope from</th>
                            <th class="vf-table__heading" scope="col">Envelope to</th>
                            <th class="vf-table__heading" scope="col">Accuracy</th>
                        </tr>
                    </thead>
                    <tbody class="vf-table__body">
                        {% for hit in funfams_data %}
                            <tr class="vf-table__row pfam-item" data-proteinfunfam="{{ hit.funfam }}">
                                <td class="vf-table__cell">
                                    <a href="{{ hit.funfam_url }}" target="_blank">{{ hit.funfam }}</a>
                                </td>
                                <td class="vf-table__cell">{{ hit.e_value|floatformat:"-g" }}</td>
                                <td class="vf-table__cell">{{ hit.score }}</td>
                                <td class="vf-table__cell">{{ hit.hmm_from }}</td>
                                <td class="vf-table__cell">{{ hit.hmm_to }}</td>
                                <td class="vf-table__cell">{{ hit.ali_from }}</td>
                                <td class="vf-table__cell">{{ hit.ali_to }}</td>
                                <td class="vf-table__cell">{{ hit.env_from }}</td>
                                <td class="vf-table__cell">{{ hit.env_to }}</td>
                                <td class="vf-table__cell">{{ hit.acc }}</td>
                            </tr>
                        {% endfor %}
                    </tbody>
                </table>
            {% else %}
                <p>No FunFam hits found</p>
            {% endif %}
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Profile-profile Pfam matches</h3>
            <div>
                <p>
                    This MGnifam HMM profile was searched against the profile Pfam database (ver. 35.0) with HHblits from HH-suite.
                </p>
            </div>
            {% if hits_data %}
                <table id="pfams-table" class="vf-table .vf-u-margin__bottom--400">
                    <thead class="vf-table__header">
                        <tr class="vf-table__row">
                            <th class="vf-table__heading" scope="col">Pfam</th>
                            <th class="vf-table__heading" scope="col">Name</th>
                            <th class="vf-table__heading" scope="col">Description</th>
                            <th class="vf-table__heading" scope="col">Probability</th>
                            <th class="vf-table__heading" scope="col">E-value</th>
                            <th class="vf-table__heading" scope="col">Length</th>
                            <th class="vf-table__heading" scope="col">MGnifam HMM</th>
                            <th class="vf-table__heading" scope="col">Pfam HMM</th>
                        </tr>
                    </thead>
                    <tbody class="vf-table__body">
                        {% for hit in hits_data %}
                            <tr class="vf-table__row pfam-item" data-proteinpfam="{{ hit.pfam }}">
                                <td class="vf-table__cell">
                                    <a href="https://www.ebi.ac.uk/interpro/entry/pfam/{{ hit.pfam }}" target="_blank">{{ hit.pfam }}</a>
                                </td>
                                <td class="vf-table__cell">{{ hit.name }}</td>
                                <td class="vf-table__cell">{{ hit.description }}</td>
                                <td class="vf-table__cell">{{ hit.prob }}</td>
                                <td class="vf-table__cell">{{ hit.e_value|floatformat:"-g" }}</td>
                                <td class="vf-table__cell">{{ hit.length }}</td>
                                <td class="vf-table__cell">{{ hit.query_hmm }}</td>
                                <td class="vf-table__cell">{{ hit.template_hmm }}</td>
                            </tr>
                        {% endfor %}
                    </tbody>
                </table>
            {% else %}
                <p>No Pfam hits found</p>
            {% endif %}
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Structure-structure hits</h3>
            <div>
                <p>
                    This MGnifam 3D structure was searched against the Alphafold/UniProt and PDB databases with foldseek.
                </p>
            </div>
            {% if structural_annotations %}
                <table id="structural-annotations-table" class="vf-table .vf-u-margin__bottom--400">
                    <thead class="vf-table__header">
                        <tr class="vf-table__row">
                            <th class="vf-table__heading" scope="col">Rank</th>
                            <th class="vf-table__heading" scope="col">Target Structure</th>
                            <th class="vf-table__heading" scope="col">Target DB</th>
                            <th class="vf-table__heading" scope="col">Aligned Length</th>
                            <th class="vf-table__heading" scope="col">Query Start</th>
                            <th class="vf-table__heading" scope="col">Query End</th>
                            <th class="vf-table__heading" scope="col">Target Start</th>
                            <th class="vf-table__heading" scope="col">Target End</th>
                            <th class="vf-table__heading" scope="col">E-value</th>
                        </tr>
                    </thead>
                    <tbody class="vf-table__body">
                        {% for annotation in structural_annotations %}
                            <tr class="vf-table__row">
                                <td class="vf-table__cell">{{ annotation.prob }}</td>
                                <td class="vf-table__cell">{{ annotation.target_structure_identifier | safe }}</td>
                                <td class="vf-table__cell">{{ annotation.target_structure_db }}</td>
                                <td class="vf-table__cell">{{ annotation.aligned_length }}</td>
                                <td class="vf-table__cell">{{ annotation.q_start }}</td>
                                <td class="vf-table__cell">{{ annotation.q_end }}</td>
                                <td class="vf-table__cell">{{ annotation.t_start }}</td>
                                <td class="vf-table__cell">{{ annotation.t_end }}</td>
                                <td class="vf-table__cell">{{ annotation.e_value|floatformat:"-g" }}</td>
                            </tr>
                        {% endfor %}
                    </tbody>
                </table>
            {% else %}
                <p>No structural matches found</p>
            {% endif %}
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400 | sequenceViewerDiv">
            <h3 class="vf-card__heading">Family Representative Sequence Viewer</h3>
            <p id="positionMessage">Amino acid position: -</p>
            <div id="proteinSequenceContainer" class="protein-sequence-container">
                {{ rep_sequence }}
            </div>
            <div>
                <p>
                    <b>HMM Consensus</b>
                    {{ consensus }}
                </p>
            </div>
            {% comment %} <div class="right-align">
                <button id="copy_button" class="vf-button vf-button--primary vf-button--sm">Copy</button>
            </div> {% endcomment %}
        </div>
    </article>

</section>

<script src="{% static 'explorer/details.js' %}"></script>
{% comment %} Biomes distribution {% endcomment %}
<script src='https://cdn.plot.ly/plotly-2.29.1.min.js'></script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.17/d3.min.js'></script>
{% comment %} PDB viewer {% endcomment %}
<script src="https://cdn.jsdelivr.net/npm/babel-polyfill/dist/polyfill.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@webcomponents/webcomponentsjs/webcomponents-lite.js" charset="utf-8"></script>
<script src="https://cdn.jsdelivr.net/npm/@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js" charset="utf-8"></script>
<script type="text/javascript" src="https://www.ebi.ac.uk/pdbe/pdb-component-library/js/pdbe-molstar-component-3.1.0.js"></script>
{% comment %} MSA Viewer {% endcomment %}
<link rel="stylesheet" href="{% static "explorer/msa/msa.min.css" %}">
<script src="{% static "explorer/msa/msa.min.js" %}"></script>
{% comment %} HMM Viewer {% endcomment %}
<link rel="stylesheet" href="{% static "explorer/hmm/hmm_logo.min.css" %}">
<script src="{% static "explorer/hmm/hmm_logo_min.js" %}"></script>

<script>
    let rf = '{{ rf }}'
    let hmm_logo_json = '{{ hmm_logo_json | safe }}'
</script>

{% comment %} Feature viewer (s4pred) - Must be inserted after HMM Viewer, else there is a hmm_logo function bug{% endcomment %}
<script src="https://cdn.jsdelivr.net/gh/calipho-sib/feature-viewer@v1.1.0/dist/feature-viewer.bundle.js"></script>

{% endblock %}
