<!-- explorer/templates/explorer/details.html -->

{% extends "explorer/base.html" %}
{% load static %}
{% block content %}
{% load custom_filters %}

<script src="{% static 'explorer/details.js' %}"></script>

<section class="vf-content | vf-stack vf-stack--600">
    <h2> MGnifam: {{ family_id }} </h2>
    <h3 style="display: inline-block; margin-right: 10px;"> Number of sequences: {{ family_size }} </h3>
    <button id="show-members-btn" class="vf-button">Toggle</button>
    <div id="family-members-div" style="display: none;">
        <span style="font-size: 0.8em;">
            {% for link in family_members_links %}
                {{ link|safe }}{% if not forloop.last %}, {% endif %}
            {% endfor %}
        </span>
    </div>
    <h3> Protein representative: <a href="http://proteins.mgnify.org/{{ protein_rep }}" target="_blank">{{ protein_rep }}</a></h3>
    <h4> Region: {{ region }} </h4>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-grid vf-grid__col-4 | vf-card__content">
            <div class="vf-grid__col--span-1">
                <h3>ESM Atlas 3D structure</h3>
                <p>
                    Predicted 3D protein structure through the Meta AI ESMFold model.
                    ESMFold uses the representations from a large language model (ESM2)
                    to generate an accurate structure prediction from the sequence of a protein.
                </p>
                <p>
                    For more information visit:
                    <ul class="vf-list">
                        <li class="vf-list__item">
                            <a href="https://esmatlas.com/" target="_blank" class="vf-link ext">ESM Metagenomic Atlas</a>
                        </li>
                        <li class="vf-list__item">
                            <a href="https://www.science.org/doi/full/10.1126/science.ade2574"
                               target="_blank"
                               class="vf-link ext">
                                Evolutionary-scale prediction of atomic-level protein structure with a language model
                            </a>
                        </li>
                    </ul>
                </p>
            </div>
            <div class="vf-grid__col--span-3">
                <div id="pdb_viewer">
                    <pdbe-molstar
                        custom-data-url="{% static "cif/"|add:cif_path %}"
                        custom-data-format="cif"
                        alphafold-view="true"
                        bg-color-r="255"
                        bg-color-g="255"
                        bg-color-b="255"
                        sequence-panel="true"
                        landscape="true"
                    >
                    </pdbe-molstar>
                </div>
            </div>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Multiple Sequence Alignment (Seed)</h3>
            <div id="msa-div"></div>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">HMM Viewer</h3>
            <div>
                <p>
                    Stack height is the information content (aka relative entropy) of the position,
                    and letters divide that height according to their estimated probability.
                    Click on a stack to highlight the relative column on the seed MSA viewer above.
                </p>
            </div>
            <div id="logo" class="logo"></div>
            <div id="col_info" class="col_info"></div>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Profile-profile comparison hits</h3>
            <div>
                <p>
                    This MGnifam HMM profile was searched against the profile Pfam database (ver. 35.0) with HHblits from HH-suite.
                </p>
            </div>
            <table id="pfams-table"
                       class="vf-table .vf-u-margin__bottom--400">
                <thead class="vf-table__header">
                    <tr class="vf-table__row">
                        <th class="vf-table__heading" scope="col">Rank</th>
                        <th class="vf-table__heading" scope="col">Pfam hit</th>
                        <th class="vf-table__heading" scope="col">MGnifam HMM</th>
                        <th class="vf-table__heading" scope="col">Pfam HMM</th>
                        <th class="vf-table__heading" scope="col">E-value</th>
                    </tr>
                </thead>
                <tbody class="vf-table__body">
                    {% for hit in hits_data %}
                        <tr class="vf-table__row pfam-item" data-proteinpfam="{{ hit.pfam_id }}">
                            <td class="vf-table__cell">{{ hit.rank }}</td>
                            <td class="vf-table__cell">
                                <a href="https://www.ebi.ac.uk/interpro/entry/pfam/{{ hit.pfam_id }}"
                                   target="_blank">{{ hit.name }}</a>
                            </td>
                            <td class="vf-table__cell">{{ hit.query_hmm }}</td>
                            <td class="vf-table__cell">{{ hit.template_hmm }}</td>
                            <td class="vf-table__cell">{{ hit.e_value }}</td>                            
                        </tr>
                    {% empty %}
                        <tr class="vf-table__heading">
                            <td class="vf-table__cell" colspan="5">No Pfam hits found</td>
                        </tr>
                    {% endfor %}
                </tbody>
            </table>
        </div>
    </article>

    <article class="vf-card vf-card--brand vf-card--bordered">
        <div class="vf-card__content | vf-stack vf-stack--400">
            <h3 class="vf-card__heading">Structural comparisons</h3>
            <div>
                <p>
                    This MGnifam 3D structure was searched against the Alphafold/UniProt, PDB and ESMAtlas30 databases with foldseek.
                </p>
            </div>
            <table id="structural-annotations-table" class="vf-table .vf-u-margin__bottom--400">
                <thead class="vf-table__header">
                    <tr class="vf-table__row">
                        <th class="vf-table__heading" scope="col">Rank</th>
                        <th class="vf-table__heading" scope="col">Target Structure</th>
                        <th class="vf-table__heading" scope="col">Aligned Length</th>
                        <th class="vf-table__heading" scope="col">Query Start</th>
                        <th class="vf-table__heading" scope="col">Query End</th>
                        <th class="vf-table__heading" scope="col">Target Start</th>
                        <th class="vf-table__heading" scope="col">Target End</th>
                        <th class="vf-table__heading" scope="col">E-value</th>
                    </tr>
                </thead>
                <tbody class="vf-table__body">
                    {% for annotation in structural_annotations %}
                        <tr class="vf-table__row">
                            <td class="vf-table__cell">{{ annotation.rank }}</td>
                            <td class="vf-table__cell">{{ annotation.target_structure_identifier | safe }}</td>
                            <td class="vf-table__cell">{{ annotation.aligned_length }}</td>
                            <td class="vf-table__cell">{{ annotation.query_start }}</td>
                            <td class="vf-table__cell">{{ annotation.query_end }}</td>
                            <td class="vf-table__cell">{{ annotation.target_start }}</td>
                            <td class="vf-table__cell">{{ annotation.target_end }}</td>
                            <td class="vf-table__cell">{{ annotation.e_value|floatformat:"-g" }}</td>
                        </tr>
                    {% empty %}
                        <tr class="vf-table__row">
                            <td class="vf-table__cell" colspan="7">No structural annotations found</td>
                        </tr>
                    {% endfor %}
                </tbody>
            </table>
        </div>
    </article>
</section>

{% comment %} Family members {% endcomment %}
<script>
    document.getElementById('show-members-btn').addEventListener('click', function() {
        var div = document.getElementById('family-members-div');
        div.style.display = div.style.display === 'none' ? 'block' : 'none';
    });
</script>

{% comment %} PDB viewer {% endcomment %}
<script src="https://cdn.jsdelivr.net/npm/babel-polyfill/dist/polyfill.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@webcomponents/webcomponentsjs/webcomponents-lite.js" charset="utf-8"></script>
<script src="https://cdn.jsdelivr.net/npm/@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js" charset="utf-8"></script>
<script type="text/javascript" src="https://www.ebi.ac.uk/pdbe/pdb-component-library/js/pdbe-molstar-component-3.1.0.js"></script>

{% comment %} MSA Viewer {% endcomment %}
<link rel="stylesheet" href="{% static "explorer/msa/msa.min.css" %}">
<script src="{% static "explorer/msa/msa.min.js" %}"></script>
<script>

    var rootDiv = document.getElementById("msa-div")

    var opts = {
        el: rootDiv,
            importURL: "../static/{{ seed_msa_filepath }}",  menu: {
            menuFontsize: "14px",
            menuItemFontsize: "14px",
            menuItemLineHeight: "14px",
            menuMarginLeft: "3px",
            menuPadding: "3px 4px 3px 4px",
        },
        bootstrapMenu: true,
        vis: {
            seqlogo: false,
            conserv: false,
            overviewbox: false,
            gapHeader: false,
            labelId: true,
            labelName: true,
            labelPartition: false,
            labelCheckbox: false
        },
        colorscheme: {
            scheme: "clustal2", // name of your color scheme
            colorBackground: true, // otherwise only the text will be colored
            showLowerCase: true, // used to hide and show lowercase chars in the overviewbox
            opacity: 0.6 //opacity for the residues
        }
    };
    var m = new msa.msa(opts);

    const translate_to_hmm_pos = (sequence, position) => {
        let hmm_pos = 0;

        for (let i = 0; i < sequence.length; i++) {
            if (sequence[i] === 'x') {
                hmm_pos = hmm_pos + 1;
                if (i >= position) break;
            }
        }

        return hmm_pos;
    };

    {% comment %} const click_hmm = (hmm_pos) => {
        const inputCol = document.querySelector('.logo_position');
        inputCol.focus();
        inputCol.value = null;
        inputCol.value = hmm_pos;
        const event = new KeyboardEvent('keypress', {
            key: 'Enter',
            code: 'Enter',
            keyCode: 13,
            which: 13,
            bubbles: true
        });
        inputCol.dispatchEvent(event);
        const inputColButton = document.querySelector('.button.logo_change');
        inputColButton.click();
    };

    m.g.on("residue:click", function(data){
        hmm_pos = translate_to_hmm_pos("{{ rf }}", data.rowPos)
        click_hmm(hmm_pos);
    }); {% endcomment %}
    m.render();
</script>

{% comment %} HMM Viewer {% endcomment %}
<link rel="stylesheet" href="{% static "explorer/hmm/hmm_logo.min.css" %}">
<script src="{% static "explorer/hmm/hmm_logo_min.js" %}"></script>
<script>
    const extract_hmm_column = (p_text) => {
        var parts = p_text.split(':');
        var numberPart = parts[parts.length - 1];
        var col = parseInt(numberPart);
        return col;
    };

    const translate_to_msa_pos = (sequence, hmm_position) => {
        let x_counter = 0;
        let msa_pos;
        for ( msa_pos = 0; msa_pos < sequence.length; msa_pos++) {
            if (sequence[msa_pos] === 'x') {
                x_counter = x_counter + 1;
                if (x_counter == hmm_position) break;
            }
        }
        return msa_pos;
    };

    const link_hmm_to_msa = () => {
        // need a dummy init p in the col_info div to tie the event
        const hiddenParagraph = document.createElement('p');
        hiddenParagraph.textContent = 'Hidden paragraph content';
        hiddenParagraph.style.display = 'none';
        const divElement = document.getElementById('col_info');
        divElement.appendChild(hiddenParagraph);
        // create the observer
        const observer = new MutationObserver(function(mutationsList, observer) {
            for(const mutation of mutationsList) {
                if (mutation.type === 'childList') {
                    mutation.addedNodes.forEach(node => {
                        if (node.tagName === 'P') {
                            const divElement = document.getElementById('col_info');
                            const pElement = divElement.querySelector('p:first-of-type');
                            let clicked_col = extract_hmm_column(pElement.textContent);
                            let msa_pos = translate_to_msa_pos("{{ rf }}", clicked_col);
                            let elements = document.getElementsByClassName('msa-col-header');
                            elements[msa_pos].click();
                        }
                    });
                }
            }
        });
        // Configure and start observing for changes in child nodes
        const config = { childList: true, subtree: true };
        observer.observe(divElement, config);
    };

    $(document).ready(function () {
        let hmmLogoJson = JSON.parse('{{ hmm_logo_json | safe }}');
        let logoDiv = document.getElementById('logo');
        logoDiv.setAttribute('data-logo', JSON.stringify(hmmLogoJson));
        $('#logo').hmm_logo({height_toggle: true, column_info: "#col_info"});
        // switch to maximum observed scale
        let radioInput = document.querySelector('input[name="scale"][value="obs"]');
        radioInput.click();
        // hide Coordinates fieldset options
        let logoSettingsDiv = document.querySelector('.logo_settings');
        let fieldsets = logoSettingsDiv.querySelectorAll('fieldset');
        fieldsets[2].style.display = 'none';

        // link event to msa
        link_hmm_to_msa();
    });
</script>

{% comment %} Datatables {% endcomment %}
<script>
    $(document).ready( function () {
        if ($('#pfams-table:contains("No Pfam hits found")').length === 0) {
            $('#pfams-table').DataTable();
        }
        if ($('#structural-annotations-table:contains("No structural annotations found")').length === 0) {
            $('#structural-annotations-table').DataTable();
        }
    } );
</script>

{% endblock %}
